libcudacxx/include/cuda/std/__format/format_spec_parser.h

```diff
@@ -0,0 +1,1199 @@
+  __ret.__alternate_form_       = true;
+  __ret.__zero_padding_         = true;
+  __ret.__locale_specific_form_ = true;
+  __ret.__type_                 = true;
+  __ret.__consume_all_          = true;
+  return __ret;
+}
+[[nodiscard]] _CCCL_API constexpr __fmt_spec_fields __fmt_spec_fields_str() noexcept
+{
+  __fmt_spec_fields __ret{};
+  __ret.__precision_   = true;
+  __ret.__type_        = true;
+  __ret.__consume_all_ = true;
+  return __ret;
+}
+[[nodiscard]] _CCCL_API constexpr __fmt_spec_fields __fmt_spec_fields_ptr() noexcept
+{
+  __fmt_spec_fields __ret{};
+  __ret.__zero_padding_ = true;
+  __ret.__type_         = true;
+  __ret.__consume_all_  = true;
+  return __ret;
+}
+
+enum class __fmt_spec_alignment : uint8_t
+{
+  // No alignment is set in the format string.
+  __default,
+  __left,
+  __center,
+  __right,
+  __zero_padding,
+};
+
+enum class __fmt_spec_sign : uint8_t
+{
+  // No sign is set in the format string.
+  //
+  // The sign isn't allowed for certain format-types. By using this value
+  // it's possible to detect whether or not the user explicitly set the sign
+  // flag. For formatting purposes it behaves the same as \ref __minus.
+  __default,
+  __minus,
+  __plus,
+  __space,
+};
+
+enum class __fmt_spec_type : uint8_t
+{
+  __default = 0,
+  __string,
+  __binary_lower_case,
+  __binary_upper_case,
+  __octal,
+  __decimal,
+  __hexadecimal_lower_case,
+  __hexadecimal_upper_case,
+  __pointer_lower_case,
+  __pointer_upper_case,
+  __char,
+  __hexfloat_lower_case,
+  __hexfloat_upper_case,
+  __scientific_lower_case,
+  __scientific_upper_case,
+  __fixed_lower_case,
+  __fixed_upper_case,
+  __general_lower_case,
+  __general_upper_case,
+};
+
+[[nodiscard]] _CCCL_API constexpr uint32_t __fmt_spec_make_type_mask(__fmt_spec_type __t)
+{
+  const auto __shift = static_cast<uint32_t>(__t);
+  if (__shift > 31)
+  {
+    _CUDA_VSTD::__throw_format_error("The type does not fit in the mask");
+  }
+  return 1 << __shift;
+}
+
+inline constexpr uint32_t __fmt_spec_type_mask_int =
+  _CUDA_VSTD::__fmt_spec_make_type_mask(__fmt_spec_type::__binary_lower_case) | //
+  _CUDA_VSTD::__fmt_spec_make_type_mask(__fmt_spec_type::__binary_upper_case) | //
+  _CUDA_VSTD::__fmt_spec_make_type_mask(__fmt_spec_type::__decimal) | //
+  _CUDA_VSTD::__fmt_spec_make_type_mask(__fmt_spec_type::__octal) | //
+  _CUDA_VSTD::__fmt_spec_make_type_mask(__fmt_spec_type::__hexadecimal_lower_case) | //
+  _CUDA_VSTD::__fmt_spec_make_type_mask(__fmt_spec_type::__hexadecimal_upper_case);
+
+// there is a bug in gcc < 10 that warns about `__fmt_spec_alignment : 3` being too small for all enum values,
+// so we use the underlying type instead
+struct __fmt_spec_std
+{
+  underlying_type_t<__fmt_spec_alignment> __alignment_ : 3;
+  underlying_type_t<__fmt_spec_sign> __sign_           : 2;
+  bool __alternate_form_                               : 1;
+  bool __locale_specific_form_                         : 1;
+  __fmt_spec_type __type_;
+};
```

Comment:
I believe we need to verify that this actually does the right thing. 

If i remember correctly, MSVC does not properly pack fields with different base types. Please verify  that this indeed packs as expected.

Also we a packing 7 bits, should we add an explicit padding here?